README.md
# Signal K client for Deno

**signalk_client** is a library to facilitate communication with a Signal K
server.

It provides the following classes to interact with the Signal K `HTTP` and
`STREAM` APIs as well as exposing STREAM `Events`:

[api](HTTP_API.md): class for interacting with Signal K HTTP API

[stream](STREAM_API.md): class for interacting with Signal K STREAM API

[apps](APPS.md): class to enable interaction with applications installed on the
Signal K server.

See [signalk-client API](#signalkclient-api) below for details.

---

## Usage

```
import { SignalKClient } from 'https://deno.land/x/signalk_client/mod.ts'
```

The simplest way to get started is to connect to a Signal K server and open a
websocket to its data stream.

To do this:

1. Use `connectStream(...)` to connect to the server and open the data stream.

2. Subscribe to the `stream` events to process the received data

_Example:_

```javascript
const signalk = new SignalKClient();

// **** Subscribe to Signal K Stream events ***

signalk.stream.events.on(
    'connect',
    (ev:Event) => {
        console.log('** Connect Event: ', ev);
    }
);

signalk.stream.events.on(
    'close',
    (ev:Event) => {
        console.log('** Close Event: ', ev);
    }
);

signalk.stream.events.on(
    'error',
    (ev:Event) => {
        console.log('** Error Event: ', ev);
    }
);

signalk.stream.events.on(
    'message',
    (ev:Event) => {
        console.log('** Message Event: ', ev);
    }
);

// **** CONNECT to Server and open STREAM ****
signalk.connectStream('192.168.99.100', 80, false, 'self');
```

Once connected you can then interact with both the STREAM and HTTP APIs in the
following ways:

- Use the `stream` object to interact with the websocket connection.

```javascript
// **** send data to STREAM API ****
signalk.stream.send({..data..});
```

- Use the `api` object to interact Signal K HTTP API path. `/signalk/v1/api/`

_Example:_

```javascript
// **** make HTTP API request ****
try {
    const response = await signalk.api.get('vessels/self/navigation/position');
}
catch {
    ...
}
```

### Connect and open Stream on Demand

---

If you want to just use the HTTP API or defer the connection to the STREAM API
based on user interaction use the `connect(..)` method.

1. Use `connect(...)` to connect to the server and perform endpoint discovery.

2. When you are ready to connect to the STREAM API use `openStream()` ( or
   `stream.open()` ) with a `null` or `undefined` _url_ parameter. This will
   cnnect to the discovered stream endpoint.

_Example:_

```javascript
const signalk = new SignalKClient();

// **** Subscribe to Signal K Stream events ***

signalk.stream.events.on(
    'connect',
    (ev:Event) => {
        console.log('** Connect Event: ', ev);
    }
);

signalk.stream.events.on(
    'close',
    (ev:Event) => {
        console.log('** Close Event: ', ev);
    }
);

signalk.stream.events.on(
    'error',
    (ev:Event) => {
        console.log('** Error Event: ', ev);
    }
);

signalk.stream.events.on(
    'message',
    (ev:Event) => {
        console.log('** Message Event: ', ev);
    }
);

// **** CONNECT to Server ****
signalk.connect('192.168.99.100', 80, false, 'self');

... 

signalk.openStream( null, 'self');

OR

signalk.stream.open( null, 'self');
```

### Use with non-HTTP enabled Signal K server

---

By default the **connect** methods will cause an HTTP request to be sent to the
server `/signalk` path to discover the server's advertised endpoints.

To interact with the server without using endpoint discovery use the
`openStream(<hostname>, <port>)` method specifying the host ip address and port.

_Note: No HTTP endpoint discovery is performed when using `openXX()` methods and
specifying a host._

_Example:_

```javascript
const signalk = new SignalKClient();

// **** Subscribe to Signal K Stream events ***

signalk.stream.events.on(
    'connect',
    (ev:Event) => {
        console.log('** Connect Event: ', ev);
    }
);

signalk.stream.events.on(
    'close',
    (ev:Event) => {
        console.log('** Close Event: ', ev);
    }
);

signalk.stream.events.on(
    'error',
    (ev:Event) => {
        console.log('** Error Event: ', ev);
    }
);

signalk.stream.events.on(
    'message',
    (ev:Event) => {
        console.log('** Message Event: ', ev);
    }
);

// **** Open the STREAM ****
signalk.openStream('192.168.99.100', 80, false, 'self');
```

# SignalKClient API

SignalKClient contains the following classes to interact with Signal K API's:

[api](HTTP_API.md): class for interacting with Signal K HTTP API

[stream](STREAM_API.md): class for interacting with Signal K STREAM API

[apps](APPS.md): class to enable interaction with applications installed on the
Signal K server.

_Follow the links for the relevant documentation._

[Attributes](#attributes)

- `server`
- `version`
- `authToken`
- `uuid`
- `signalkUuid`
- `proxied`

[Methods](#methods)

- `hello()`
- `connect()`
- `disconnect()`
- `connectStream()`
- `connectPlayback()`
- `openStream()`
- `openPlayback()`
- `snapshot()`
- `accessRequest()`
- `checkAccessRequest()`
- `get()`
- `put()`
- `post()`
- `login()`
- `logout()`
- `isLoggedIn()`
- `setAppId()`
- `setAppVersion()`
- `appDataVersions()`
- `appDataKeys()`
- `appDataGet()`
- `appDataSet()`
- `appDataPatch()`

---

### Attributes:

`server`:

Information returned from Signal K server _hello_ response.

```javascript
{
    endpoints: {},
    info: {},
    apiVersions: []
}
```

_apiVersions_ Constains list of api versions supported by the Signal K server.

---

`version`:

Get / Set preferred Signal K API version to use when the server supports more
than one version. _This must be used prior to connecting to the server._ If the
Signal K server does not supports the specified version `v1` will be used.

_Note: Signal K API is currently only available in `v1`._

_Example:_

```javascript
    // ** set target version **
    signalk.version=2;

    // ** connect to server **
    signalk.connect(...);

    // ** get the version currently in use **
    console.log(signalk.version)
```

---

`authToken`:

A token string to be used for authentication when interacting with the Signal K
server.

Use the `login()` method to authenticate to the server and retrieve a token for
the specified user.

_Example:_

```javascript
signalk.authToken = "<auth_token_string>";
```

Once you have supplied an `authToken` it will be used for all subsequent
operations.

---

`uuid`:

Returns a v4 UUID string

_Example:_

```javascript
let uuid = signalk.uuid;

// returns 27b88354-9fe0-4952-9ce6-c9d4eaea6d9e
```

---

`signalkUuid`:

Returns a formatted Signal K resource identifier

_Example:_

```javascript
let uuid = signalk.signalkUuid;

// returns urn:mrn:signalk:uuid:27b88354-9fe0-4952-9ce6-c9d4eaea6d9e
```

---

`proxied`:

Boolean value to indicate whether the Signal K server that is being connected to
is behind a proxy server.

- `false` (default): Uses endpoints received in the `hello()` response.

- `true`: Replaces protocol, host & port values of endpoint values received in
  the `hello()` response with those from connection.

_Example: `proxied = false` (default)_

```javascript
proxied = false;

// Signal K server url: http://myServer.org:3000
signalk.connect('myServer.org');

// hello response
{
    ...
    endpoints: {
        "signalk-http":"http://myServer.org:3000/signalk/v1/api/",
        "signalk-ws":"ws://myServer.org:3000/signalk/v1/stream"
    }
} 

// endpoints used are those received in hello response.
```

_Example: `proxied = true`_

```javascript
proxied = true;

// Proxied Signal K server url: https://myServer.org:3100
signalk.connect('myServer.org', 3100, true);

//hello response 
{
    ...
    endpoints: {
        "signalk-http":"http://myServer.org:3000/signalk/v1/api/",
        "signalk-ws":"ws://myServer.org:3000/signalk/v1/stream"
    }
} 

// received endpoint values are modified to include the proxy url values.
endpoints: {
    "signalk-http":"https://myServer.org:3100/signalk/v1/api/",
    "signalk-ws":"wss://myServer.org:3100/signalk/v1/stream"
}
```

---

### Methods:

`hello(hostname, port, useSSL): Promise`

Send _discovery_ request to the Signal K server `/signalk` path.

_Note: SignalKClient is not considered **connected** after using this method._

_Parameters:_

- _hostname_: host name or ip address

- _port_: port number

- _useSSL_: true: uses secure socket protocols _(https / wss)_

_Returns_: Observable

_Example:_

```javascript
// **** make Discovery request ****
const response = await signalk.hello("myServer", 80, false);
```

---

`connect(hostname, port, useSSL, subscribe): Promise`

This method performs the following:

1. Issues a `hello()` request
2. Populates the `server` attibute with the received data

_Parameters:_

- _hostname_: host name or ip address

- _port_: port number

- _useSSL_: true: uses secure socket protocols _(https / wss)_

- _subscribe_: Signal K subcription request value: 'all', 'self' or 'none'.
  _(Uses server default if null)_

_Returns_: Promise

_Example:_

```javascript
const response = await signalk.connect("myServer", 80, false);
```

---

`disconnect()`

Disconnects from Signal K server and closes all connections.

_Example:_

```javascript
signalk.disconnect();
```

---

`connectStream(hostname, port, useSSL, subscribe): Promise<boolean>`

Connect to Signal K server and and open a connection to the STREAM API after
performing service endpoint discovery.

This method performs the following:

1. Calls `connect()`
2. Opens a connection to the discovered Stream endpoint.

_Parameters:_

- _hostname_: host name or ip address

- _port_: port number

- _useSSL_: true: uses secure socket protocols _(https / wss)_

- _subscribe_: Signal K subcription request value: 'all', 'self' or 'none'.
  _(Uses server default if null)_

_Returns_: Promise

_Example:_

```javascript
// **** Subscribe to Signal K Stream events ***

signalk.stream.events.on(
    'connect',
    (ev:Event) => {
        console.log('** Connect Event: ', ev);
    }
);

signalk.stream.events.on(
    'close',
    (ev:Event) => {
        console.log('** Close Event: ', ev);
    }
);

signalk.stream.events.on(
    'error',
    (ev:Event) => {
        console.log('** Error Event: ', ev);
    }
);

signalk.stream.events.on(
    'message',
    (ev:Event) => {
        console.log('** Message Event: ', ev);
    }
);

// **** CONNECT to Delta Stream ****
signalk.connectStream('192.168.99.100', 80, false, 'self');
```

---

`connectPlayback(hostname, port, useSSL, options): Promise<boolean>`

Connect to Signal K server and and open a connection to the PLAYBACK STREAM API
after performing service endpoint discovery.

This method performs the following:

1. Calls `connect()`
2. Calls `openPlayback()`.

_Parameters:_

- _hostname_: host name or ip address

- _port_: port number

- _useSSL_: true: uses secure socket protocols _(https / wss)_

- _options_: Signal K Playback options

```javascript
{ 
    startTime: *Date / Time to start playback from.
    playbackRate: A number defining the rate at which data is sent.
    subscribe: 'all', 'self' or 'none'. *(Uses server default if null)*
}
```

_Returns_: Promise

_Example:_

```javascript
// **** Subscribe to Signal K Stream events ***

signalk.stream.events.on(
    'connect',
    (ev:Event) => {
        console.log('** Connect Event: ', ev);
    }
);

signalk.stream.events.on(
    'close',
    (ev:Event) => {
        console.log('** Close Event: ', ev);
    }
);

signalk.stream.events.on(
    'error',
    (ev:Event) => {
        console.log('** Error Event: ', ev);
    }
);

signalk.stream.events.on(
    'message',
    (ev:Event) => {
        console.log('** Message Event: ', ev);
    }
);  

// **** CONNECT to Playback Stream ****

const response = await signalk.connectPlayback('myServer', 80, false, {
    subscribe: 'self',
    playbackRate: 1,
    startTime: '2019-01-19T07:14:58Z'
});
```

---

`openStream(url, subscribe, token)`

Connect direct to Signal K server DELTA stream using the supplied parameters
without performing _endpoint discovery_.

This method is for use when there is no HTTP API available.

_Paramaters:_

- _url_: url of Signal K stream endpoint.

- _subscribe_: Signal K subcription request value: 'all', 'self' or 'none'.
  _(Uses server default if null)_

- _token_: Authentication token.


CHANGELOG.md
# CHANGELOG: signalk-client

### v1.0.1

- add `validate()` method for token validation / renewal

### v1.0.0

Initial Release


signalk-client/signalk-client.ts
import { SignalKHttp } from "./http-api.ts";
import { SignalKStream } from "./stream-api.ts";
import { SignalKApps } from "./apps-api.ts";
import { Message, Path } from "./utils.ts";
import { debug } from "./mod.ts";

interface Server_Info {
  endpoints: { [key: string]: EndPoint };
  info: SKServer;
  apiVersions: Array<string>;
}

interface EndPoint {
  version: string;
  "signalk-http": string;
  "signalk-ws": string;
}

/** Signal K Server information */
export interface SKServer {
  version: string;
  id: string;
}

interface HelloResponse {
  endpoints: { [key: string]: EndPoint };
  server: SKServer;
}

interface PlaybackOptions {
  startTime: string;
  playbackRate: number;
  subscribe: string | undefined;
}

interface JSON_Patch {
  op: "add" | "replace" | "remove" | "copy" | "move" | "test";
  path: string;
  value: unknown;
}

/** Application data contexts */
export enum APPDATA_CONTEXT {
  USER = "user",
  GLOBAL = "global",
}

/** Signal K Client class */
export class SignalKClient {
  /** Signal K server hostname / ip address.
   * @private
   */
  private hostname = "localhost";
  /** Port to connect on.
   * @private
   */
  private port = 3000;
  /** Protocol to use http / https.
   * @private
   */
  private protocol = "";

  /** Signal K API version to use
   * @private
   */
  private _version = "v1";

  /** Authentication token value
   * @private
   */
  private _token = ""; // token for when security is enabled on the server

  /** endpoints to fallback to if hello response is not received.
   * @private
   */
  private fallbackEndpoints: HelloResponse = {
    endpoints: {
      v1: {
        version: "1.0.0",
        "signalk-http": "",
        "signalk-ws": "",
      },
    },
    server: { id: "fallback", version: "1.43.0" },
  };

  /** Server information block */
  public server: Server_Info = {
    endpoints: {},
    info: {
      version: "",
      id: "",
    },
    apiVersions: [],
  };

  /** Endpoints fallback to host address when no hello response */
  public fallback = false;

  /** endpoints are set to host address regardless of contents of hello response */
  public proxied = false;

  /** get preferred Signal K API version to use
   * @property
   */
  get version(): number {
    return parseInt(this._version.slice(1));
  }

  /** set preferred Signal K API version to use
   * @param val Major version number of API. e.g. 1
   */
  set version(val: number) {
    const v: string = "v" + val;
    if (this.server.apiVersions.length === 0) {
      this._version = v;
      debug(`Signal K api version set to: ${v}`);
    } else {
      this._version = this.server.apiVersions.includes(v) ? v : this._version;
      debug(
        `Signal K api version set request: ${v}, result: ${this._version}`,
      );
    }
    this.api.version = parseInt(v.slice(1));
    this.stream.version = parseInt(v.slice(1));
  }
  /** set auth token value
   * @param val Token value
   */
  set authToken(val: string) {
    this._token = val;
    this.api.authToken = val;
    this.stream.authToken = val;
  }

  /** Get a Message object */
  get message(): Message {
    return Message;
  }

  /** Generate and return a v4 UUID */
  get uuid(): string {
    return crypto.randomUUID(); //new UUID();
  }

  /** Generate and return a Signal K UUID */
  get signalkUuid(): string {
    return `urn:mrn:signalk:uuid:${crypto.randomUUID()}`;
  }

  /** Signal K Apps API */
  public apps: SignalKApps;
  /** Signal K HTTP API */
  public api: SignalKHttp;
  /** Signal K stream API */
  public stream: SignalKStream;

  /** Create new Signal K Client */
  constructor() {
    this.apps = new SignalKApps();
    this.api = new SignalKHttp();
    this.stream = new SignalKStream();
    this.init();
  }

  /** Initialise client protocol, hostname and port values.
   * @param hostname Signal K server hostname / IP address
   * @param port Port on which Signal K server is listening
   * @param useSSL If true uses https / wss, if false uses http /ws
   */
  private init(
    hostname: string = this.hostname,
    port = 80,
    useSSL = false,
  ) {
    this.hostname = hostname ? hostname : this.hostname;
    if (useSSL) {
      this.protocol = "https";
      this.port = port || 443;
    } else {
      this.protocol = "http";
      this.port = port || 80;
    }
    const httpUrl = `${this.protocol}://${this.hostname}:${this.port}`;
    const wsUrl = `${useSSL ? "wss" : "ws"}://${this.hostname}:${this.port}`;
    this.fallbackEndpoints.endpoints.v1[
      "signalk-http"
    ] = `${httpUrl}/signalk/v1/api/`;
    this.fallbackEndpoints.endpoints.v1[
      "signalk-ws"
    ] = `${wsUrl}/signalk/v1/stream`;
  }

  // **************** CONNECTION AND DISCOVERY  ********************

  /** Signal K server endpoint discovery request (/signalk).
   * @param hostname Signal K server hostname / IP address
   * @param port Port on which Signal K server is listening
   * @param useSSL If true uses https / wss, if false uses http /ws
   */
  hello(
    hostname: string = this.hostname,
    port = 3000,
    useSSL = false,
  ): Promise<{ [key: string]: unknown }> {
    this.init(hostname, port, useSSL);
    return this.get("/signalk");
  }

  /** connect to server (endpoint discovery) and DO NOT open Stream
   * @param hostname Signal K server hostname / IP address
   * @param port Port on which Signal K server is listening
   * @param {useSSL If true uses https / wss, if false uses http /ws
   */
  async connect(
    hostname: string = this.hostname,
    port = 3000,
    useSSL = false,
  ): Promise<boolean> {
    debug("Contacting Signal K server.........");
    try {
      const response: unknown = await this.hello(hostname, port, useSSL);
      // ** discover endpoints **
      this.getLoginStatus();
      if (this.stream) {
        this.stream.close();
      }
      this.processHello(response as HelloResponse);
      this.api.endpoint = this.resolveHttpEndpoint();
      this.stream.endpoint = this.resolveStreamEndpoint();
      return true;
    } catch (error) {
      if (this.fallback) {
        // fallback if no hello response
        if (this.stream) {
          this.stream.close();
        }
        this.processHello();
        this.api.endpoint = this.resolveHttpEndpoint();
        this.stream.endpoint = this.resolveStreamEndpoint();
        return true;
      } else {
        this.disconnectedFromServer();
        throw error;
      }
    }
  }

  /** Close all connections to Signal K server */
  disconnect() {
    this.stream.close();
  }

  /** Connect + open Delta Stream (endpoint discovery)
   * @param hostname Signal K server hostname / IP address
   * @param port Port on which Signal K server is listening
   * @param useSSL If true uses https / wss, if false uses http /ws
   * @param subscribe Subscription parameters for stream connection
   */
  connectStream(
    hostname: string = this.hostname,
    port = 3000,
    useSSL = false,
    subscribe = "",
  ): Promise<boolean> {
    return new Promise((resolve, reject) => {
      this.connect(hostname, port, useSSL)
        .then(() => {
          // ** connect to stream api at preferred version else fall back to default version
          const url = this.resolveStreamEndpoint();
          if (!url) {
            reject(false);
            return;
          }
          this.stream.open(url, subscribe);
          resolve(true);
        })
        .catch((e) => {
          throw e;
        });
    });
  }

  /** connect to playback stream (endpoint discovery)
   * @param hostname Signal K server hostname / IP address
   * @param port Port on which Signal K server is listening
   * @param useSSL If true uses https / wss, if false uses http /ws
   * @param options Options for playback stream connection
   */
  connectPlayback(
    hostname: string = this.hostname,
    port = 3000,
    useSSL = false,
    options: PlaybackOptions,
  ): Promise<boolean> {
    return new Promise((resolve) => {
      this.connect(hostname, port, useSSL)
        .then(() => {
          // ** connect to playback api at preferred version else fall back to default version
          this.openPlayback("", options, this._token);
          resolve(true);
        })
        .catch((e) => {
          throw e;
        });
    });
  }

  /** connect to delta stream with (NO endpoint discovery)
   * @param url Signal K stream endpoint url
   * @param subscribe Subscription parameters for stream connection
   * @param token Authentication token
   */
  openStream(
    url: string = this.hostname,
    subscribe?: string,
    token?: string,
  ): boolean {
    debug("openStream.........");
    if (!url) {
      // connect to stream api at discovered endpoint
      url = this.resolveStreamEndpoint();
      if (!url) {
        throw new Error("Server has no advertised Stream endpoints!");
      }
    }
    this.stream.open(url, subscribe, token);
    return true;
  }

  /** connect to playback stream (NO endpoint discovery)
   * @param url Signal K stream endpoint url
   * @param options Options for playback stream connection
   * @param token Authentication token
   */
  openPlayback(
    url: string = this.hostname,
    options?: PlaybackOptions,
    token?: string,
  ): boolean {
    debug("openPlayback.........");
    if (!url) {
      // connect to stream api at discovered endpoint
      url = this.resolveStreamEndpoint();
      if (!url) {
        throw new Error("Server has no advertised Stream endpoints!");
      }
      url = url.replace("stream", "playback");
    }
    let pb = "";
    let subscribe: string | undefined = "";
    if (options && typeof options === "object") {
      pb += options.startTime
        ? "?startTime=" +
          options.startTime.slice(0, options.startTime.indexOf(".")) +
          "Z"
        : "";
      pb += options.playbackRate ? `&playbackRate=${options.playbackRate}` : "";
      subscribe = options.subscribe ? options.subscribe : undefined;
    }
    this.stream.open(url + pb, subscribe, token);
    return true;
  }

  /** process Hello response
   * @params Recieved hello response from server.
   */
  private processHello(response?: HelloResponse) {
    if (this.proxied) {
      this.server.endpoints = this.fallbackEndpoints.endpoints;
    } else {
      this.server.endpoints = response && response["endpoints"]
        ? response["endpoints"]
        : this.fallbackEndpoints.endpoints;
    }
    this.server.info = response && response["server"]
      ? response["server"]
      : this.fallbackEndpoints.server;
    this.server.apiVersions = this.server.endpoints
      ? Object.keys(this.server.endpoints)
      : [];
    debug(this.server.endpoints);
    this.api.server = this.server.info;
    this.apps.endpoint = this.resolveAppsEndpoint();
  }

  /** Return signalk apps API url */
  private resolveAppsEndpoint(): string {
    return this.resolveHttpEndpoint().replace("api", "apps");
  }

  /** Return preferred WS stream url */
  private resolveStreamEndpoint(): string {
    if (
      this.server.endpoints[this._version] &&
      this.server.endpoints[this._version]["signalk-ws"]
    ) {
      debug(`Connecting endpoint version: ${this._version}`);
      return `${this.server.endpoints[this._version]["signalk-ws"]}`;
    } else if (
      this.server.endpoints["v1"] &&
      this.server.endpoints["v1"]["signalk-ws"]
    ) {
      debug(`Connection falling back to: v1`);
      return `${this.server.endpoints["v1"]["signalk-ws"]}`;
    } else {
      return "";
    }
  }

  /** Return signalk-http endpoint url */
  private resolveHttpEndpoint(): string {
    let url = "";
    if (this.server.endpoints[this._version]) {
      // ** connection made
      // ** connect to http endpoint at prescribed version else fall back to default version
      if (this.server.endpoints[this._version]["signalk-http"]) {
        url = `${this.server.endpoints[this._version]["signalk-http"]}`;
      } else {
        url = `${this.server.endpoints["v1"]["signalk-http"]}`;
      }
    } else {
      const msg =
        "No current connection http endpoint service! Use connect() to establish a connection.";
      debug(msg);
    }
    return url;
  }

  /** Cleanup after server disconnection. */
  private disconnectedFromServer() {
    this.server.endpoints = {};
    this.server.info = {
      version: "",
      id: "",
    };
    this.server.apiVersions = [];
  }

  /** HTTP GET from API path.
   * @param path Signal K path.
   */
  async get(path: string): Promise<{ [key: string]: unknown }> {
    if (path && path.length > 0 && path[0] === "/") {
      path = path.slice(1);
    }
    const url = `${this.protocol}://${this.hostname}:${this.port}/${
      Path.dotToSlash(path)
    }`;

    debug(`get ${url}`);
    const options: RequestInit = {};

    if (this._token) {
      options.headers = new Headers({ Authorization: `JWT ${this._token}` });
    }
    const response = await fetch(url, options);
    return response.json();
  }

  /** HTTP PUT to API path
   * @param path Signal K path.
   * @param value Value to apply.
   */
  async put(path: string, value: unknown): Promise<Response> {
    const url = `${this.protocol}://${this.hostname}:${this.port}/${
      Path.dotToSlash(path)
    }`;

    debug(`put ${url}`);
    const headers = new Headers({ "Content-Type": "application/json" });
    if (this._token) {
      headers.append("Authorization", `JWT ${this._token}`);
    }

    const options: RequestInit = {
      method: "PUT",
      headers: headers,
      body: JSON.stringify(value),
    };
    return await fetch(url, options);
  }

  /** HTTP POST to API path.
   * @param {string} path Signal K path.
   * @param value Value to apply.
   */
  async post(path: string, value: unknown): Promise<Response> {
    if (path && path.length > 0 && path[0] === "/") {
      path = path.slice(1);
    }
    const url = `${this.protocol}://${this.hostname}:${this.port}/${
      Path.dotToSlash(path)
    }`;

    debug(`post ${url}`);
    const headers = new Headers({ "Content-Type": "application/json" });
    if (this._token) {
      headers.append("Authorization", `JWT ${this._token}`);
    }

    const options: RequestInit = {
      method: "POST",
      headers: headers,
      body: JSON.stringify(value),
    };
    return await fetch(url, options);
  }

  /** Login and retrieve an  auth token for supplied user details
   * @param username User id
   * @param password User password
   */
  async login(
    username: string,
    password: string,
  ): Promise<{ ok: boolean; status: number; token: string }> {
    const headers = new Headers({ "Content-Type": "application/json" });
    const url =
      `${this.protocol}://${this.hostname}:${this.port}/signalk/${this._version}/auth/login`;

    debug(`post ${url}`);

    if (this._token) {
      headers.append("Authorization", `JWT ${this._token}`);
    }
    const options = {
      method: "POST",
      headers: headers,
      body: JSON.stringify({ username: username, password: password }),
    };

    const response = await fetch(
      url,
      options,
    );

    const result = {
      ok: response.ok,
      status: response.status,
      token: "",
    };
    if (response.ok) {
      const cookies = response.headers.get("set-cookie")?.split(";");
      cookies?.forEach((c) => {
        if (c.indexOf("JAUTHENTICATION") !== -1) {
          this.authToken = c.split("=")[1];
          debug("authToken: ", this._token);
          result.token = this._token;
        }
      });
    }
    return result;
  }

  /** Validate / refresh token from server */
  async validate(): Promise<{ ok: boolean; status: number; token: string }> {
    const url =
      `${this.protocol}://${this.hostname}:${this.port}/signalk/${this._version}/auth/validate`;

    debug(`post ${url}`);
    const headers = new Headers({ "Content-Type": "application/json" });
    if (this._token) {
      headers.append("Authorization", `JWT ${this._token}`);
    }

    const options = {
      method: "POST",
      headers: headers,
      body: null,
    };

    const response: Response = await fetch(url, options);

    const result = {
      ok: response.ok,
      status: response.status,
      token: "",
    };
    if (response.ok) {
      const cookies = response.headers.get("set-cookie")?.split(";");
      cookies?.forEach((c) => {
        if (c.indexOf("JAUTHENTICATION") !== -1) {
          this.authToken = c.split("=")[1];
          debug("authToken: ", this._token);
          result.token = this._token;
        }
      });
    }
    return result;
  }

  /** Logout from server */
  logout(): Promise<boolean> {
    const url =
      `${this.protocol}://${this.hostname}:${this.port}/signalk/${this._version}/auth/logout`;

    debug(`post ${url}`);
    const headers = new Headers({ "Content-Type": "application/json" });
    if (this._token) {
      headers.append("Authorization", `JWT ${this._token}`);
    }

    const options = {
      method: "POST",
      headers: headers,
      body: null,
    };

    try {
      fetch(url, options);
      return Promise.resolve(true);
    } catch {
      return Promise.resolve(false);
    }
  }

  /** Tests if a user authenticated to the server
   * @returns True if user is logged in.
   */
  async isLoggedIn(): Promise<boolean> {
    const response = await this.getLoginStatus();
    return response.status === "loggedIn" ? true : false;
  }

  /** Fetch Signal K login status from server */
  getLoginStatus(): Promise<{ [key: string]: unknown }> {
    let url = `/skServer/loginStatus`;
    if (this.server && this.server.info.id === "signalk-server-node") {
      const ver = this.server.info["version"].split(".");
      url = ver[0] === "1" && parseInt(ver[1]) < 36 //use legacy link for older versions
        ? `/loginstatus`
        : url;
    }
    return this.get(url);
  }

  /** Get data via the snapshot http api path for supplied time.
   * @param Signal K context
   * @param time As ISO formatted time string.
   */
  snapshot(context: string, time: string): Promise<{ [key: string]: unknown }> {
    if (!time) {
      throw new Error("Error: No time value supplied!");
    }
    time = time.slice(0, time.indexOf(".")) + "Z";
    let url = this.resolveHttpEndpoint();
    if (!url) {
      throw new Error("Error: Unable to resolve URL!");
    }
    url = `${url.replace("api", "snapshot")}${
      Path.contextToPath(
        context,
      )
    }?time=${time}`;

    return this.get(url);
  }

  /*******************************
   *  Access Requests
   *******************************/
  public clientId: string | undefined;

  /** Access request methods
   * @param name Name of sensor / process requesting access
   * @param id Client  id
   */
  async accessRequest(
    name: string,
    id?: string,
  ): Promise<{ [key: string]: unknown }> {
    if (!name) {
      throw new Error("Error: Name not supplied!");
    }

    if (!id && !this.clientId) {
      this.clientId = this.uuid;
    } else if (!this.clientId) {
      this.clientId = id;
    }

    const response = await this.post(
      `/signalk/${this._version}/access/requests`,
      {
        clientId: this.clientId,
        description: name,
      },
    );
    return response.json();
  }

  /** Check status of an access request.
   * @param href url path returned in the access request PENDING response.
   */
  async checkAccessRequest(href: string): Promise<{ [key: string]: unknown }> {
    if (!href) {
      throw new Error("Error: href not supplied!");
    }
    return await this.get(href);
  }

  /*******************************
   *  applicationData api methods
   * context: 'user' or 'global'
   * appId: application id string
   *******************************/

  /** Target application id.
   * @private
   */
  private _appId = "";

  /** Target application version.
   * @private
   */
  private _appVersion = "";

  /** Return endpoint for target app. */
  private resolveAppDataEndpoint(
    context: APPDATA_CONTEXT,
    appId: string,
  ): string {
    if (!context || !appId) {
      return "";
    }
    const url = this.resolveHttpEndpoint().replace("api", "applicationData");
    return `${url}${context}/${appId}/`;
  }

  /** Set the appId.
   * @param value Application id.
   */
  setAppId(value: string) {
    this._appId = value;
  }

  /** Set the app version.
   * @param value Application version.
   */
  setAppVersion(value: string) {
    this._appVersion = value;
  }

  /** Get list of available versions of app data stored
   * @param context Signal K context
   * @param appId Application id.
   */
  appDataVersions(
    context: APPDATA_CONTEXT = APPDATA_CONTEXT.USER,
    appId: string = this._appId,
  ): Promise<{ [key: string]: unknown }> {
    const url = this.resolveAppDataEndpoint(context, appId);
    return this.get(url);
  }

  /** Get list of available keys for a stored path
   * @param path Path to app data
   * @param context Signal K context
   * @param appId Application id
   * @param version Version of data
   */
  appDataKeys(
    path = "",
    context: APPDATA_CONTEXT = APPDATA_CONTEXT.USER,
    appId: string = this._appId,
    version: string = this._appVersion,
  ): Promise<{ [key: string]: unknown }> {
    path = path.length != 0 && path[0] === "/" ? path.slice(1) : path;
    let url = this.resolveAppDataEndpoint(context, appId);
    url += `${version}/${path}?keys=true`;
    return this.get(url);
  }

  /** Get stored value at provided path
   * @param path Path to app data
   * @param context Signal K context
   * @param appId Application id
   * @param version Version of data
   */
  appDataGet(
    path = "",
    context: APPDATA_CONTEXT = APPDATA_CONTEXT.USER,
    appId: string = this._appId,
    version: string = this._appVersion,
  ): Promise<{ [key: string]: unknown }> {
    path = path.length != 0 && path[0] === "/" ? path.slice(1) : path;
    let url = this.resolveAppDataEndpoint(context, appId);
    url += `${version}/${path}`;
    return this.get(url);
  }

  /** Set stored value at provided path
   * @param path Path to app data
   * @param context Signal K context
   * @param appId Application id
   * @param version Version of data
   */
  appDataSet(
    path: string,
    value: { [key: string]: unknown },
    context: APPDATA_CONTEXT = APPDATA_CONTEXT.USER,
    appId: string = this._appId,
    version: string = this._appVersion,
  ): Promise<Response> {
    if (!path) {
      throw new Error("Error: Invalid path!");
    }
    if (path[0] === "/") {
      path = path.slice(1);
    }
    const ep = this.resolveAppDataEndpoint(context, appId);
    if (!ep) {
      throw new Error("Error: Invalid path!");
    }
    return this.post(
      `${ep}${version}/${Path.dotToSlash(path)}`,
      value,
    );
  }

  /** update / patch stored values using Array of JSON patch objects
   * @param path Path to app data
   * @param context Signal K context
   * @param appId Application id
   * @param version Version of data
   */
  appDataPatch(
    value: Array<JSON_Patch>,
    context: APPDATA_CONTEXT = APPDATA_CONTEXT.USER,
    appId: string = this._appId,
    version: string = this._appVersion,
  ): Promise<Response> {
    const ep = this.resolveAppDataEndpoint(context, appId);
    if (!ep || !version) {
      throw new Error("Error: Invalid path or version!");
    }
    return this.post(`${ep}${version}`, value);
  }
}
